internal/core/cloud/aws.go

```diff
@@ -151,11 +173,45 @@ func requestMetadata(ctx context.Context, client HTTPClient, url string) (interf
 		return nil, err
 	}
 
+	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
+		return nil, errors.Errorf("failed to fetch AWS metadata: %s", resp.Status)
+	}
+
 	// The metadata endpoint may return json elements
 	if json.Valid(body) {
-		var jsonData interface{}
+		var jsonData any
 		err := json.Unmarshal(body, &jsonData)
 		return jsonData, err
 	}
 	return string(body), nil
 }
+
+func requestMetadataToken(ctx context.Context, client HTTPClient) (string, error) {
+	log.Debug("Fetching IMDS token...")
+
+	url := fmt.Sprintf("%sapi/token", awsMetadataURL)
+
+	req, _ := http.NewRequestWithContext(ctx, http.MethodPut, url, nil)
+
+	req.Header.Add(metadataTokenTTLHeader, strconv.Itoa(awsEC2MetadataTokenTTL))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Debugf("An error occurred while fetching IMDS token: %s", err)
+		return "", err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.Errorf("failed to fetch metadata token: %s", resp.Status)
+	}
+
+	defer resp.Body.Close()
```

Comment:
nitpick: should be maybe run this `defer` before the previous `if` conditions?

Reply:
Shouldn't be necessary, should it? 
I mean that if we return in previous if conditions, the body is not read at all, so we don't need defering closing it. Am I missing something?

Reply:
I always thought that we needed to close it if we get a not errored response from the client queries, but I haven't checked more than that

Reply:
I found this for example: https://stackoverflow.com/questions/33238518/what-could-happen-if-i-dont-close-response-body
```
If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close.
```
I would at least use the defer before doing the forbidden/authorized check